### ISSUE TITLE
MySQL8: Client does not support authentication protocol requested by server

### ISSUE THREAD

**Issue by benderillo** [strapi/strapi#13774]
### Required System information

- Node.js version: 16
- Strapi version: 4.2.2
- Database: MySQL 8
- Operating system: Mac

### Describe the bug

**TL;DR**
Mysql database driver used by Strapi does not support the new authentication method that MySQL8 uses by default.
This generates error:
```
Error: ER_NOT_SUPPORTED_AUTH_MODE: Client does not support authentication protocol requested by server; consider upgrading MySQL client
```

**Long story**
Currently Strapi uses `mysql` npm package as a driver to connect to MySQL database. This package does not support the new authentication method (`caching_sha2_password `) which some MySQL8 hosted databases use by default. 
For example, if you deploy on Railway.app, their mysql db will refuse connection.

The solution can be either lower the security of the Database by configuring it to use the old `mysql_native_password ` auth method or use another package for mysql driver which supports the authentication: [mysql2](https://www.npmjs.com/package/mysql2)
The details can be found in this answer in stackoverflow: https://stackoverflow.com/a/56509065/985942

Unfortunately, using the `mysql2` package as a driver for the DB fails because strapi passes the client name ('mysql') as a package name to "knex" so the "knex" fails to find such a package upon loading. 
Specifying a "mysql2" as a client in the config file fails too as the Strapi database package has explicit checks for "mysql" client name.

This ticket is to ask to add support for the `mysql2` driver to strapi so developers can opt to install it instead of the `mysql`.
Since the `mysql2` package is mostly API compatible with the `mysqljs` it should be an easy fix to treat the driver `mysql2` the same way as the `mysql`.

### Steps to reproduce the behaviour

1. Create a new strapi project and configure it to use MySQL database
2. Install a MySQL8 database and configure project to connect to it.
3. Run Strapi project and observe it fails to connect because of the  `ER_NOT_SUPPORTED_AUTH_MODE` error.
4. Install mysql2 `npm install mysql2` and change the database configuration to use "mysql2" driver.
5. Try to run Strapi and observe that it fails as the driver is not supported.

### Expected behaviour

Should be able to select "mysql2" as a driver for a MySQL database.

**UPDATE**
PR for this issue: https://github.com/strapi/strapi/pull/13775

**Comment by derrickmehaffy** [comment]
Not really a bug but related to: https://feedback.strapi.io/developer-experience/p/add-support-for-mysql2-mysql-v8-auth-system

**Comment by SunSparc** [comment]
MySQL version 8 was released 4 years ago. Seems like it is probably time to update the authentication methods to match.

**Comment by derrickmehaffy** [comment]
> MySQL version 8 was released 4 years ago. Seems like it is probably time to update the authentication methods to match.

Please feel free to submit a PR :) Else for the moment it's not really a priority as the simple work around is using the `native_password` authentication method in MySQL 8

**Comment by AndreyKraynoff** [comment]
made database.js like this:

module.exports = ({ env }) => ({
  connection: {
    client: 'mysql2',
    connection: {
      host: env('DATABASE_HOST', '127.0.0.1'),
      port: env.int('DATABASE_PORT', 3306),
      database: env('DATABASE_NAME', 'DATABASE_NAME'),
      user: env('DATABASE_USERNAME', 'DATABASE_USERNAME'),
      password: env('DATABASE_PASSWORD', 'DATABASE_PASSWORD'),
      ssl: env.bool('DATABASE_SSL', true),
    },
  },
});
 
And got this error:  Unknown dialect mysql2

**Comment by ricky11** [comment]
any update this? strangely it was working , but either railway (MySQL host) or strapi had some changes that broke when I try to run npm run develop.

**Comment by derrickmehaffy** [comment]
Not currently, we haven't had time yet to draft a solid PR

**Comment by mouhsnimohamed** [comment]
the same error  for me, any help?

**Comment by mtt786** [comment]
I have solved this using a custom user other than the root.
```mysql
ALTER USER 'customuser'@'localhost' IDENTIFIED WITH mysql_native_password BY 'yourpassword';
FLUSH PRIVILEGES;
```

**Comment by mouhsnimohamed** [comment]
> I have solved this using a custom user other than the root.
> 
> ```sql
> ALTER USER 'customuser'@'localhost' IDENTIFIED WITH mysql_native_password BY 'yourpassword';
> FLUSH PRIVILEGES;
> ```

it worked for me

**Comment by ivosh** [comment]
We have a requirement to authenticate against MySQL 8 database with a client X.509 certificate. That is, *not* using a password.
Obviously this implies also using TLS protocol. Thus using `mysql_native_password` authentication mechanism is not an option.

I tried to `hack` in `mysql2` support, now that strapi uses knex v1, by simply copying directory `@strapi/database/lib/dialects/mysql` to `@strapi/database/lib/dialects/mysql2` and adding the following two lines into
`@strapi/database/lib/dialects/index.js`:
```
case 'mysql2':
      return require('./mysql2');
```

At this point, typical strapi operations work fine.

**Comment by yuesth** [comment]
I have the same issue about using MySQL 8.x  on Strapi 4.x, you can easily reference this https://strapi.io/blog/configuring-strapi-mysql-database to fix the problem. 

Basically, the problem caused by your auth method MySQL is still using `cache_sha2_password` plugin which is not supported by Strapi client, instead, you should downgrade by using `mysql_native_password` plugin.

**Comment by uguratakan** [comment]
> I have solved this using a custom user other than the root.
> 
> ```sql
> ALTER USER 'customuser'@'localhost' IDENTIFIED WITH mysql_native_password BY 'yourpassword';
> FLUSH PRIVILEGES;
> ```

Worked for me too

**Comment by Khaan25** [comment]
You have to check “Use Legacy Password Encryption” in your MySQL installation. Not the version 8 one

**Comment by friday-sds** [comment]
This is one is stll not reflecting with the latest changes 
My package.json looks like this 
`  "scripts": {
    "develop": "strapi develop",
    "start": "strapi start",
    "build": "strapi build",
    "strapi": "strapi"
  },
  "dependencies": {
    "@strapi/plugin-i18n": "4.9.0",
    "@strapi/plugin-users-permissions": "4.9.0",
    "@strapi/strapi": "4.9.0",
    "better-sqlite3": "8.0.1",
  },`
  
  Adding mysql2 is giving unknown dialect error

**Comment by derrickmehaffy** [comment]
> @innerdvations  This is one is stll not reflecting with the latest changes 
> My package.json looks like this 
> `  "scripts": {
>     "develop": "strapi develop",
>     "start": "strapi start",
>     "build": "strapi build",
>     "strapi": "strapi"
>   },
>   "dependencies": {
>     "@strapi/plugin-i18n": "4.9.0",
>     "@strapi/plugin-users-permissions": "4.9.0",
>     "@strapi/strapi": "4.9.0",
>     "better-sqlite3": "8.0.1",
>   },`
>   
>   Adding mysql2 is giving unknown dialect error

The PR was merged but hasn't been released yet, if you look at the PR it's milestone is set to v4.9.1 not v4.9.0

**Comment by friday-sds** [comment]
> > @innerdvations  This is one is stll not reflecting with the latest changes
> > My package.json looks like this
> > `  "scripts": { "develop": "strapi develop", "start": "strapi start", "build": "strapi build", "strapi": "strapi" }, "dependencies": { "@strapi/plugin-i18n": "4.9.0", "@strapi/plugin-users-permissions": "4.9.0", "@strapi/strapi": "4.9.0", "better-sqlite3": "8.0.1", },`
> > Adding mysql2 is giving unknown dialect error
> 
> The PR was merged but hasn't been released yet, if you look at the PR it's milestone is set to v4.9.1 not v4.9.0

Acknowledged !! will wait for release !! anyway just realised that planetscale doesn't work with Strapi !!

**Comment by derrickmehaffy** [comment]
> > > @innerdvations  This is one is stll not reflecting with the latest changes
> > > My package.json looks like this
> > > `  "scripts": { "develop": "strapi develop", "start": "strapi start", "build": "strapi build", "strapi": "strapi" }, "dependencies": { "@strapi/plugin-i18n": "4.9.0", "@strapi/plugin-users-permissions": "4.9.0", "@strapi/strapi": "4.9.0", "better-sqlite3": "8.0.1", },`
> > > Adding mysql2 is giving unknown dialect error
> > 
> > The PR was merged but hasn't been released yet, if you look at the PR it's milestone is set to v4.9.1 not v4.9.0
> 
> Acknowledged !! will wait for release !! anyway just realised that planetscale doesn't work with Strapi !!

Indeed I believe they have issues with foreign key constraints or something along those lines which we rely on heavily for all databases.

**Comment by WDevigne** [comment]
you can use this solution:

oper windows CMD
Login to your MySQL server by typing: mysql -u root -p
then type: ALTER USER 'your_username'@'localhost' IDENTIFIED WITH 'mysql_native_password' BY 'your_password';
then: FLUSH PRIVILEGES;
finally: exit;
back to your project terminal and try again: npm run develop
👍

**Comment by axeloz** [comment]
mysql2 works fine with Strapi. 
I issued a `npm install mysql2 --save` in the package.json.
Then `DATABASE_CLIENT=mysql2` (or update the config/database.js file)
Works now.

**Comment by Aleksandr-Zharkov** [comment]
> mysql2 works fine with Strapi. I issued a `npm install mysql2 --save` in the package.json. Then `DATABASE_CLIENT=mysql2` (or update the config/database.js file) Works now.

Thank you for your advise. I can confirm that migrating from `mysql` to `mysql2` works with Strapi 4.10.7 version. And yes, don't forget to change it to mysql2 in your env. variables (and restart everything)

**Comment by bayuly94** [comment]
> mysql2 works fine with Strapi. I issued a `npm install mysql2 --save` in the package.json. Then `DATABASE_CLIENT=mysql2` (or update the config/database.js file) Works now.

thanks

**Comment by BigKongfuPanda** [comment]
> mysql2 works fine with Strapi. I issued a `npm install mysql2 --save` in the package.json. Then `DATABASE_CLIENT=mysql2` (or update the config/database.js file) Works now.

thanks, worked for me

**Comment by prateekarya100** [comment]
ALTER USER 'your_db_username'@'localhost' IDENTIFIED WITH mysql_native_password BY 'your_db_password';
FLUSH PRIVILEGES;

**Comment by kriskuiper** [comment]
> mysql2 works fine with Strapi. I issued a `npm install mysql2 --save` in the package.json. Then `DATABASE_CLIENT=mysql2` (or update the config/database.js file) Works now.

Thanks! Can confirm that this one did the trick 👌🏻

**Comment by fossil31** [comment]
Execute the following query in MYSQL Workbench

ALTER USER 'root'@'localhost' IDENTIFIED WITH mysql_native_password BY 'password';

Where root as your user localhost as your URL and password as your password

Then run this query to refresh privileges:

flush privileges;

Try connecting using node after you do so.

If that doesn't work, try it without @'localhost' part.

**Comment by IvanAdmaers** [comment]
> mysql2 works fine with Strapi. I issued a `npm install mysql2 --save` in the package.json. Then `DATABASE_CLIENT=mysql2` (or update the config/database.js file) Works now.

Works for me too. Strapi v4.19.0

**Comment by wahyunurarizky-sera** [comment]
> We have a requirement to authenticate against MySQL 8 database with a client X.509 certificate. That is, _not_ using a password. Obviously this implies also using TLS protocol. Thus using `mysql_native_password` authentication mechanism is not an option.
> 
> I tried to `hack` in `mysql2` support, now that strapi uses knex v1, by simply copying directory `@strapi/database/lib/dialects/mysql` to `@strapi/database/lib/dialects/mysql2` and adding the following two lines into `@strapi/database/lib/dialects/index.js`:
> 
> ```
> case 'mysql2':
>       return require('./mysql2');
> ```
> 
> At this point, typical strapi operations work fine.

nice
